controllers/topology/internal/extensions/patches/engine_test.go

```diff
@@ -0,0 +1,541 @@
+	"sigs.k8s.io/cluster-api/controllers/topology/internal/scope"
+	"sigs.k8s.io/cluster-api/internal/builder"
+	. "sigs.k8s.io/cluster-api/internal/matchers"
+)
+
+func TestApply(t *testing.T) {
+	type patch struct {
+		name    string
+		patches []api.GenerateResponsePatch
+	}
+	type expectedFields struct {
+		infrastructureCluster                          map[string]interface{}
+		controlPlane                                   map[string]interface{}
+		controlPlaneInfrastructureMachineTemplate      map[string]interface{}
+		machineDeploymentBootstrapTemplate             map[string]map[string]interface{}
+		machineDeploymentInfrastructureMachineTemplate map[string]map[string]interface{}
+	}
+
+	tests := []struct {
+		name           string
+		patches        []patch
+		expectedFields expectedFields
+	}{
+		{
+			name: "Should preserve desired state, if there are no patches",
+		},
+		{
+			name: "Should apply JSON patches to InfraCluster, ControlPlane and ControlPlaneInfrastructureMachineTemplate",
+			patches: []patch{
+				{
+					name: "fake-patch1",
+					patches: []api.GenerateResponsePatch{
+						{
+							TemplateID: api.TemplateID{
+								APIVersion:   builder.InfrastructureGroupVersion.String(),
+								Kind:         builder.GenericInfrastructureClusterTemplateKind,
+								TemplateType: api.InfrastructureClusterTemplateTemplateType,
+							},
+							Patch: apiextensionsv1.JSON{Raw: []byte(`
+[{"op":"add","path":"/spec/template/spec/resource","value":"infraCluster"}]
+							`)},
+							PatchType: api.JSONPatchType,
+						},
+						{
+							TemplateID: api.TemplateID{
+								APIVersion:   builder.ControlPlaneGroupVersion.String(),
+								Kind:         builder.GenericControlPlaneTemplateKind,
+								TemplateType: api.ControlPlaneTemplateTemplateType,
+							},
+							Patch: apiextensionsv1.JSON{Raw: []byte(`
+[{"op":"add","path":"/spec/template/spec/resource","value":"controlPlane"}]
+							`)},
+							PatchType: api.JSONPatchType,
+						},
+						{
+							TemplateID: api.TemplateID{
+								APIVersion:   builder.InfrastructureGroupVersion.String(),
+								Kind:         builder.GenericInfrastructureMachineTemplateKind,
+								TemplateType: api.ControlPlaneInfrastructureMachineTemplateTemplateType,
+							},
+							Patch: apiextensionsv1.JSON{Raw: []byte(`
+[{"op":"add","path":"/spec/template/spec/resource","value":"controlPlaneInfrastructureMachineTemplate"}]
+							`)},
+							PatchType: api.JSONPatchType,
+						},
+					},
+				}},
+			expectedFields: expectedFields{
+				infrastructureCluster: map[string]interface{}{
+					"spec.resource": "infraCluster",
+				},
+				controlPlane: map[string]interface{}{
+					"spec.resource": "controlPlane",
+				},
+				controlPlaneInfrastructureMachineTemplate: map[string]interface{}{
+					"spec.template.spec.resource": "controlPlaneInfrastructureMachineTemplate",
+				},
+			},
+		},
+		{
+			name: "Should apply JSON patches to MachineDeployment templates",
```

Comment:
Q: should we have a patch that apply to all machineDeployments?

Reply:
I added a test case to patch the InfraMachineTemplates of both MDs with the same patch (via two individual patches).

Did you mean that or did you mean I should use a patch which has TemplateType "MachineDeployment/InfrastructureMachineTemplate" and has no MachineDeployment.TopologyName and MachineDeployment.Class set and should thus apply to both?
(That's currently not supported because currently the templateID of the response must be equal to the one of the request and we're always setting those two fields for MD templates)

Reply:
I was thinking to a patch without MachineDeployment.TopologyName and MachineDeployment.Class, thus avoiding the users to define via two or more individual patches to apply the same change.
But I'm also fine to open an issue and address this improvement down the line

Reply:
In this case those fields are filled on the Request by the Patch Engine on the Response by the json patch generator. (or later on potentially by external patches)

The selector on the patch struct which the user fills out requires to specify MachineDeployment names to match MachineDeployments, but we wanted to have that explicit.

I think it would be good to keep it strict for now and later on when we define the patch extension API we should nail down if we want to support matching all MDs in that case or only explicitly matched ones (explicit matching would be inline with how the selector in the ClusterClass behaves)